frontend/src/common/hooks/useCategoryFilter.ts

```diff
@@ -462,15 +593,93 @@ function isFilterEqual<T extends Categories>(
   );
 }
 
+/**
+ * Determine if the given selected value is a selected category value key (and not a range).
+ * @param selectedValue - Selected filter value, either a category value key (e.g. "normal") or a range (e.g. [0, 10]).
+ * @returns True if given selected value is a selected category value.
+ */
+function isCategoryValueKey(
+  selectedValue: CategoryValueKey | Range
+): selectedValue is CategoryValueKey {
+  return !Array.isArray(selectedValue);
+}
+
+/**
+ * Determine if the given category value is a select category value (and not a range category value).
+ * @param categorySetValue - Range or category set value.
+ * @returns True if category set value is a set of category value keys.
+ */
+function isCategorySetCategoryKeyValue(
+  categorySetValue: CategorySetValue
+): categorySetValue is Set<CategoryValueKey> {
+  return (categorySetValue as Set<CategoryValueKey>).has !== undefined;
+}
+
+/**
+ * Determine if the given category value is a select category value (and not a range category value).
+ * @param categoryValue - Range or select category value.
+ * @returns True if category value is a select category value.
+ */
+function isSelectCategoryValue(
+  categoryValue: RangeCategory | KeyedSelectCategoryValue
+): categoryValue is KeyedSelectCategoryValue {
+  return (categoryValue as KeyedSelectCategoryValue).get !== undefined;
+}
+
+/**
+ * Update selected state of categories to match the current set of selected filters.
+ * @param nextFilterState - Filter state being built on select of filter.
+ * @param filters - Current set of selected category values (values) or ranges keyed by category (id).
+ */
+function setSelectedStates<T extends Categories>(
```

Comment:
Refactored (and renamed) [flagSelectedCategoryValues()](https://github.com/chanzuckerberg/single-cell-data-portal/pull/2017/files#diff-e6d68019c2a59230f504bd6a5d662355f67c0d10c25fadc0823bfc27d7815d5aL387) to handle range functionality.